### DISCUSSION TITLE
How do you implement Vercel's SSL in the code for email?  Error: self-signed certificate in certificate chain

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#74036] [Help] [answered]
### Summary

I recently change from Microsoft 365 Business Basics to Microsoft 365 Business Standard, and now  contact forms in my website don't work anymore.

The code I have been using is:

```
export async function POST(req) {
  const transporter = nodemailer.createTransport({
    service: "Outlook365",
    auth: {
      user: process.env.EMAIL,
      pass: process.env.EPASSWORD,
    },
    
  });
```
^^ This was working awesome, until recently, now I get an error:

```
⨯ Error: self-signed certificate in certificate chain
    at TLSSocket.onConnectSecure (node:_tls_wrap:1679:34)
    at TLSSocket.emit (node:events:524:28)
    at TLSSocket._finishInit (node:_tls_wrap:1078:8)
    at ssl.onhandshakedone (node:_tls_wrap:864:12)
    at TLSWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
  code: 'ESOCKET',
  command: 'CONN'
}
 POST /my/email-test 500 in 120ms
```

If I add this after user and pass it works, but this should be only for development.

```
tls: {
      rejectUnauthorized: false, // Allow self-signed certificates (development only)
    },
```

So?  What do I do for production??? Apparently it should be something like this?  but it isn't working:

```
export async function POST(req) {
  const transporter = nodemailer.createTransport({
    service: "Outlook365",
    auth: {
      user: process.env.EMAIL,
      pass: process.env.EPASSWORD,
    },
    
    tls: {
      rejectUnauthorized: true, 
      // ca: fs.readFileSync('path/to/your/self-signed-cert.pem'),  
      // since I don't have a file I copy and post it the SSL code given by Vercel.
      ca: process.env.SSLCERTIFICATE,
    },
  });

```

I got the SSL from here:
![image](https://github.com/user-attachments/assets/f4724fd9-b97b-4240-b6ce-fade93f1a7f0)

**Comment by icyJoseph** [comment]
Where did you get the certificate? from letsencrypt or something like that?

**Reply by Elindo586** [reply]
No, that is the SSL certificate from Vercel.

Look under: 
https://vercel.com/account/domains/tu.biz

**Comment by Elindo586** [comment]
Has anybody else have problems with this?

I am trying to fix it with Microsoft, but their phone people basically only say to buy a new certificate from a trusted source, however, my emails were working just fine until recently with new Microsoft 365 Business Standard.

Are the SSL from Vercel not good enough for Microsoft Business Standard package?

**Comment by Elindo586** [comment]
Further testing shows that the Vercel SSL works just fine at production with the live websites.

During development I need to disable security to test it, else it gives me an error that didn't give me before.

I still need to test my second email as it still gives an auth error during development, yet it works fine in production.

**Reply by icyJoseph** [reply]
Yeah, I mean Vercel SSL is not self signed or anything.

My guess is that this is a problem with the email server that's receiving your request. I found a thread in nodemailer but it was quite long, and I had little time available

**Reply by icyJoseph** [reply]
Here's one for example: https://github.com/nodemailer/nodemailer/issues/1429

**Reply by Elindo586** [reply]
So far it is working on live website to send contact form information. (So it does send emails over the web)

It does give errors during production if you don't set the rejectunauthorized to false.

Now my localhost is not SSL by Vercel, so sending and using Vercel SSL from my local host is a different story... I might do a thread about it...

**Reply by Elindo586** [reply]
Orrrr... wait...

Are you saying Vercel is not a self-sign SSL????

Like OMG...  How the heck I can send contact form emails to my emails over the web?... that is interesting..

Also... I was sending automatic emails from my localhost until recently.

Now, if I do automatic emails over a web hosting it by Vercel... then the emails might not be good?

Maybe I need to buy a self-sign certificate somewhere and make it work with Vercel for my contact forms, as well as make it work from my localhost.

**Reply by Elindo586** [reply]
This is from Chat GTP.

is vercel ssl good for emails to avoid self-sign error?
ChatGPT said:
ChatGPT
Vercel provides SSL certificates for securing web traffic to your applications, ensuring encrypted connections through HTTPS. However, when it comes to email communications, SSL certificates are a bit more nuanced.

SSL for Email:
For emails, SSL certificates are often used to secure the connection between mail servers (e.g., when sending email via SMTP or receiving it via IMAP/POP3). However, the SSL certificates used by Vercel are for securing web traffic (HTTPS) for domains hosted on their platform, not for email.

Key Points:
Vercel SSL for Websites: Vercel's SSL certificates (provided through Let's Encrypt) are primarily for your website's domain to ensure secure HTTPS connections. This SSL is for web traffic, not email traffic.

Email SSL Certificates: For emails, you need to ensure that the sending (SMTP) and receiving (IMAP/POP3) servers have proper SSL/TLS certificates installed. If you're using a custom domain for email, ensure that:

The mail server (like Gmail, Mailgun, or other providers) has a valid SSL/TLS certificate.
Your email client (like Outlook or Thunderbird) trusts the email server's certificate.
Self-Signed Certificates: A common error like "self-signed certificate" usually occurs when the email server uses a certificate that isn't issued by a trusted Certificate Authority (CA). This can be problematic for email clients or servers trying to verify the server's authenticity.

If your email server is using a self-signed certificate, you'd need to either replace it with a certificate from a trusted CA or configure the client to accept the self-signed certificate (though this is not recommended for security reasons).
Conclusion:
Vercel SSL certificates are great for HTTPS connections for websites, but they do not apply to email communications directly. To avoid self-signed certificate errors for email:

Make sure your email provider uses a valid SSL/TLS certificate.
If you’re running your own email server, ensure it’s configured with a proper SSL certificate from a trusted CA, not a self-signed one.
Use services like Mailgun, SendGrid, or Google Workspace, which handle the email security for you and avoid the self-signed certificate issue.

**Reply by icyJoseph** [reply]
Nodemailer is running on the Server side.

**Reply by Elindo586** [reply]
I use Nodemailer to send, but how does it come into play with the certificate?

I am currently sending email OK using Nodemailer on the live web for Contact Forms in my site.

I get errors when I try to send emails from my localhost.

Sooo...  should I buy a new certificate the live site? Maybe I need it to send emails from my localhost, but what if I send automatic emails using a live site hosted by Vercel?

I am still not clear why I am able to send the emails on the live site and not get errors... also wonder if those emails might get flagged as not secure.

**Reply by Elindo586** [reply]
Anyway... I must get going today, but I'll drop this here just FYI....  I am still not 100% clear of what I should do, but I will continue another day.

can I use vercel SSL to send emails using Nodemailer and Outlook without getting a self-sign error or flag?
ChatGPT said:
ChatGPT
Yes, you can use Vercel's SSL when sending emails using Nodemailer with Outlook without encountering a self-signed certificate error or flag. However, there are a few things to keep in mind to ensure it works smoothly:

1. Nodemailer with SSL/TLS
Nodemailer uses SSL/TLS to securely send emails. When using Vercel, SSL certificates should already be set up for your server, meaning you can send emails over a secure connection, assuming you configure it properly. This means that Vercel’s own SSL certificate (provided for your deployed app) ensures that the connection is secure.

2. Outlook SMTP Server Configuration
Outlook's SMTP server uses TLS (Transport Layer Security) by default, so you must ensure that you set up Nodemailer to use the correct SMTP settings for Outlook (Microsoft 365 or Office 365), which are:

SMTP server: smtp-mail.outlook.com
Port: 587 (for TLS/STARTTLS)
TLS: True (use STARTTLS, which is the proper way to start the secure connection)
3. Nodemailer Configuration Example
Here is a sample Nodemailer configuration to use with Outlook:

js
Copy code
const nodemailer = require("nodemailer");

const transporter = nodemailer.createTransport({
  service: 'hotmail', // or 'outlook'
  host: 'smtp-mail.outlook.com',
  port: 587,
  secure: false, // true for 465, false for other ports (use false for 587 and STARTTLS)
  auth: {
    user: 'your-email@outlook.com',
    pass: 'your-email-password' // It's recommended to use an app-specific password if 2FA is enabled
  },
  tls: {
    rejectUnauthorized: false // This disables self-signed certificate rejection if necessary
  }
});

const mailOptions = {
  from: 'your-email@outlook.com',
  to: 'recipient@example.com',
  subject: 'Test Email',
  text: 'Hello from Nodemailer!'
};

transporter.sendMail(mailOptions, (error, info) => {
  if (error) {
    console.log("Error: ", error);
  } else {
    console.log("Message sent: ", info.response);
  }
});
4. Avoiding Self-Signed Certificate Errors
Vercel SSL: Vercel automatically provides SSL for custom domains through Let's Encrypt, which is a trusted authority, so it should not cause any SSL issues unless you're using a local development server or incorrect SSL certificates.
Outlook's SSL Certificate: Outlook uses a valid and trusted SSL certificate, so no self-signed errors should occur when connecting to smtp-mail.outlook.com.
If you're still facing issues related to self-signed certificates, it's likely due to either:

Misconfiguration on your local machine's certificate store.
A misconfigured TLS setting in your code.
Issues with how the email service (Outlook) is responding or how Vercel's SSL interacts with outbound connections.
5. Test Deployment
Ensure you test the configuration by deploying your application on Vercel, since local setups might behave differently in terms of SSL verification. If everything is configured correctly, Vercel’s SSL should not interfere with your email sending process.

Key Notes:
Port 587 (with secure: false and tls: { rejectUnauthorized: false }) is usually the best choice for Outlook as it uses STARTTLS.
If you're using 2FA with your Outlook account, create an App Password instead of using your main account password for authentication.
In conclusion, Vercel’s SSL should not interfere with Nodemailer when connecting to Outlook as long as you use the correct port and settings. If you continue to face self-signed certificate issues, they are more likely related to your local environment or configuration rather than the Vercel or Outlook setup.

**Comment by Elindo586** [comment, marked as answer]
The short answer for this is that SSL from Vercel works to send emails on the live website, and outlook doesn't seem to have a problem while you send a contact form information via the web using Node and Vercel.

The problem many people see is while testing it from your localhost, because your localhost is not an SSL hosting like your website on the live app hosted by Vercel.

Seems however, I was able to send a bunch of emails from my localhost before, but still not very good because the receiver might flag it as insecure, and Microsoft apparently may not like to be sending frequent emails from a non SSL server.
